File: src/test/scala/com/abaddon83/burraco/shares/games/GameIdentityTest.scala
package com.abaddon83.burraco.shares.games

import java.util.UUID

import org.scalatest.funsuite.AnyFunSuite

class GameIdentityTest extends AnyFunSuite{

  test("new GameIdentity using UUID"){
    val expectedUUID = UUID.randomUUID()
    val identity = GameIdentity(expectedUUID)
    assert(identity.isInstanceOf[GameIdentity])
    assert(identity.convertTo() == expectedUUID)
  }

  test("new GameIdentity using a valid UUID String"){
    val expectedUUID = UUID.randomUUID()
    val identity = GameIdentity(expectedUUID.toString)
    assert(identity.isInstanceOf[GameIdentity])
    assert(identity.convertTo() == expectedUUID)
  }

  test("new GameIdentity using a not valid UUID String should fail"){

    assertThrows[AssertionError]{
      GameIdentity("fake-UUID")
    }
  }

}

File: src/test/kotlin/com/abaddon83/burracoGame/writeModel/models/games/GameIdentityTest.kt
package com.abaddon83.burracoGame.writeModel.models.games

import com.abaddon83.burracoGame.writeModel.models.players.PlayerIdentity
import com.abaddon83.burracoGame.writeModel.models.players.PlayerIdentityCustomSerializer
import kotlinx.serialization.json.Json
import org.junit.Test
import java.util.*
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith

class GameIdentityTest {
    @Test
    fun `new GameIdentity using UUID`(){
        val expectedUUID = UUID.randomUUID()
        val identity = GameIdentity(expectedUUID)
        assert(identity.convertTo() == expectedUUID)
    }
    @Test
    fun `new GameIdentity using a valid UUID String`(){
        val expectedUUID = UUID.randomUUID()
        val identity = GameIdentity.create(expectedUUID.toString())
        assert(identity.convertTo() == expectedUUID)
    }
    @Test
    fun `new GameIdentity using a not valid UUID String should fail`(){

        assertFailsWith(IllegalArgumentException::class){
            GameIdentity.create("fake-UUID")
        }
    }

    @Test
    fun `given a GameIdentity when I serialise it, then I should have the same GameIdentity deserialized`() {
        val expectedUUID = UUID.randomUUID()
        val identity = GameIdentity.create(expectedUUID.toString())

        val jsonString = Json.encodeToString(GameIdentityCustomSerializer,identity);
        val deserializedTris = Json.decodeFromString<GameIdentity>(GameIdentityCustomSerializer,jsonString)
        assertEquals(identity.id,deserializedTris.id)

    }
}